opentelemetry-otlp/src/retry.rs

```diff
@@ -0,0 +1,495 @@
+    pub initial_delay_ms: u64,
+    /// Maximum delay in milliseconds between retries.
+    pub max_delay_ms: u64,
+    /// Maximum jitter in milliseconds to add to the delay.
+    pub jitter_ms: u64,
+}
+
+/// A runtime stub for when experimental_async_runtime is not enabled.
+/// This allows retry policy to be configured but no actual retries occur.
+#[cfg(not(any(
+    feature = "experimental-grpc-retry",
+    feature = "experimental-http-retry"
+)))]
+#[derive(Debug, Clone, Default)]
+pub struct NoOpRuntime;
+
+#[cfg(not(any(
+    feature = "experimental-grpc-retry",
+    feature = "experimental-http-retry"
+)))]
+impl NoOpRuntime {
+    /// Creates a new no-op runtime.
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+// Generates a random jitter value up to max_jitter
+#[cfg(any(
+    feature = "experimental-grpc-retry",
+    feature = "experimental-http-retry"
+))]
+fn generate_jitter(max_jitter: u64) -> u64 {
+    let nanos = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .subsec_nanos();
+
+    let mut hasher = DefaultHasher::default();
+    hasher.write_u32(nanos);
+    hasher.finish() % (max_jitter + 1)
+}
+
+/// Retries the given operation with exponential backoff, jitter, and error classification.
+///
+/// This function provides sophisticated retry behavior by classifying errors
+/// and honoring server-provided throttling hints (e.g., HTTP Retry-After, gRPC RetryInfo).
+///
+/// # Arguments
+///
+/// * `runtime` - The async runtime to use for delays.
+/// * `policy` - The retry policy configuration.
+/// * `error_classifier` - Function to classify errors for retry decisions.
+/// * `operation_name` - The name of the operation being retried.
+/// * `operation` - The operation to be retried.
+///
+/// # Returns
+///
+/// A `Result` containing the operation's result or an error if max retries are reached
+/// or a non-retryable error occurs.
+#[cfg(any(
+    feature = "experimental-grpc-retry",
+    feature = "experimental-http-retry"
+))]
+pub async fn retry_with_backoff<R, F, Fut, T, E, C>(
+    runtime: R,
+    policy: RetryPolicy,
+    error_classifier: C,
+    operation_name: &str,
+    mut operation: F,
+) -> Result<T, E>
+where
+    R: Runtime,
+    F: FnMut() -> Fut,
+    E: std::fmt::Debug,
+    Fut: Future<Output = Result<T, E>>,
+    C: Fn(&E) -> RetryErrorType,
+{
+    let mut attempt = 0;
+    let mut delay = policy.initial_delay_ms;
+
+    loop {
+        match operation().await {
+            Ok(result) => return Ok(result), // Return the result if the operation succeeds
+            Err(err) => {
+                // Classify the error
+                let error_type = error_classifier(&err);
+
+                match error_type {
+                    RetryErrorType::NonRetryable => {
+                        otel_warn!(name: "OtlpRetry", message = format!("Operation {:?} failed with non-retryable error: {:?}", operation_name, err));
```

Comment:
Let's stick with structured logging as much as possible instead of stringifying. i.e Operation and Error should be own fields.